--- dhcpcd-3.0.17/pathnames.h.orig	2007-04-14 09:54:09.000000000 +0200
+++ dhcpcd-3.0.17/pathnames.h	2007-04-14 17:04:15.418888566 +0200
@@ -33,8 +33,8 @@
 #define NISSERVICE          ETCDIR "/init.d/ypbind"
 #define NISRESTARTARGS      "--quiet", "conditionalrestart"
 
-#define NTPFILE             ETCDIR "/ntp.conf"
-#define NTPDRIFTFILE        ETCDIR "/ntp.drift"
+#define NTPFILE             ETCDIR "/ntp/ntp.conf"
+#define NTPDRIFTFILE        ETCDIR "/ntp/drift"
 #define NTPLOGFILE          "/var/log/ntp.log"
 #define NTPSERVICE          ETCDIR "/init.d/ntpd"
 #define NTPRESTARTARGS      "--quiet", "conditionalrestart"
