--- pathnames.h	2006-11-27 21:45:26.000000000 +0100
+++ pathnames.h-new	2006-12-06 05:49:54.000000000 +0100
@@ -27,8 +27,8 @@
 #define ETCDIR			"/etc"
 #define RESOLVFILE		ETCDIR "/resolv.conf"
 #define NISFILE			ETCDIR "/yp.conf"
-#define NTPFILE			ETCDIR "/ntp.conf"
-#define NTPDRIFTFILE		ETCDIR "/ntp.drift"
+#define NTPFILE			ETCDIR "/ntp/conf"
+#define NTPDRIFTFILE		ETCDIR "/ntp/drift"
 #define DEFAULT_SCRIPT		ETCDIR "/" PACKAGE ".sh"
 
 #define STATEDIR		"/var"
