--- ./dhcpconfig.c.org3	Mon Jan  6 12:36:42 2003
+++ ./dhcpconfig.c	Mon Jan  6 12:39:21 2003
@@ -381,18 +381,14 @@
     * because the resolver won't notice the change in resolv.conf */
       (void)res_init();
     }
-  if ( ReplNISConf )
+  if ( ReplNISConf && DhcpOptions.len[nisDomainName] )
     {
       yp_renamed=1+rename(NIS_CONF,""NIS_CONF".sv");
       f=fopen(NIS_CONF,"w");
       if ( f )
 	{
 	  int i;
-	  char *domain=NULL;
-	  if ( DhcpOptions.len[nisDomainName] )
-	    domain=(char *)DhcpOptions.val[nisDomainName];
-	  else
-	    domain=(char *)DhcpOptions.val[domainName];
+	  char *domain=(char *)DhcpOptions.val[nisDomainName]; setdomainname(domain, strlen(domain)+1);
 	  for (i=0;i<DhcpOptions.len[nisServers];i+=4)
 	    fprintf(f,"domain %s server %u.%u.%u.%u\n",(domain?domain:"localdomain"),
 	    ((unsigned char *)DhcpOptions.val[nisServers])[i],
