--- dhcpcd-1.3.19-pl2/configure.in~	Thu Oct 12 03:29:08 2000
+++ dhcpcd-1.3.19-pl2/configure.in	Wed Dec 13 18:37:02 2000
@@ -39,23 +39,6 @@
 ADDS+= -DOLD_LINUX_VERSION
 fi
 
-case $OSNAME in
-	Linux)	CFLAGS=" -O2 -fomit-frame-pointer -Wall $ADDS"
-		;;
-	*)	CFLAGS=
-esac
-
-case $ARCH in
-	i486)	CFLAGS="$CFLAGS -m486 -malign-loops=2 -malign-jumps=2 -malign-functions=2"
-		;;
-	i[356]86)	CFLAGS=$CFLAGS
-		;;
-	*)	CFLAGS=$CFLAGS
-		;;
-esac
-
-LDFLAGS=' -s -O2'
-
 dnl
 dnl Checks for programs.
 dnl
