--- dhcpcd-3.1.8/config.h.orig	2007-11-17 14:11:04.000000000 +0100
+++ dhcpcd-3.1.8/config.h	2007-11-26 20:14:13.718144517 +0100
@@ -42,15 +42,15 @@
 #define PACKAGE             "dhcpcd"
 
 #define ETCDIR              "/etc"
-#define RESOLVFILE          ETCDIR "/resolv.conf"
+#define RESOLVFILE          "/etc/resolv.conf"
 
-#define NISFILE             ETCDIR "/yp.conf"
+#define NISFILE             "/etc/yp.conf"
 
-#define NTPFILE             ETCDIR "/ntp.conf"
-#define NTPDRIFTFILE        ETCDIR "/ntp.drift"
+#define NTPFILE             "/etc/ntp/ntp.conf"
+#define NTPDRIFTFILE        "/etc/ntp/ntp.drift"
 #define NTPLOGFILE          "/var/log/ntp.log"
 
-#define OPENNTPFILE         ETCDIR "/ntpd.conf"
+#define OPENNTPFILE         "/etc/ntpd/ntpd.conf"
 
 #define DEFAULT_SCRIPT      ETCDIR "/" PACKAGE ".sh"
 
@@ -70,9 +70,9 @@
  * SYSV should cover everthing else */
 #ifdef ENABLE_OPENRC
 # define SERVICE             "OPENRC"
-# define NISSERVICE          ETCDIR "/init.d/ypbind"
+# define NISSERVICE          "/etc/rc.d/init.d/ypbind"
 # define NISRESTARTARGS      "--nodeps", "--quiet", "conditionalrestart"
-# define NTPSERVICE          ETCDIR "/init.d/ntpd"
+# define NTPSERVICE          "/etc/rc.d/init.d/ntpd"
 # define NTPRESTARTARGS      "--nodeps", "--quiet", "conditionalrestart"
 #elif ENABLE_BSDRC
 # define SERVICE             "BSDRC"
