--- dhcpcd-3.1.5/config.h	2007-08-27 22:43:51.614826538 +0300
+++ dhcpcd-3.1.5/config.h	2007-08-27 23:23:49.078758408 +0300
@@ -52,21 +52,21 @@
 #define RESOLVCONF          "/sbin/resolvconf"
 
 #define ETCDIR              "/etc"
-#define RESOLVFILE          ETCDIR "/resolv.conf"
+#define RESOLVFILE          "/etc/resolv.conf"
 
-#define NISFILE             ETCDIR "/yp.conf"
-#define NISSERVICE          ETCDIR "/init.d/ypbind"
-#define NISRESTARTARGS      "--nodeps", "--quiet", "conditionalrestart"
+#define NISFILE             "/etc/yp.conf"
+#define NISSERVICE          "/etc/rc.d/init.d/ypbind"
+#define NISRESTARTARGS      "condrestart"
 
-#define NTPFILE             ETCDIR "/ntp.conf"
-#define NTPDRIFTFILE        ETCDIR "/ntp.drift"
+#define NTPFILE             "/etc/ntp/ntp.conf"
+#define NTPDRIFTFILE        "/etc/ntp/ntp.drift"
 #define NTPLOGFILE          "/var/log/ntp.log"
-#define NTPSERVICE          ETCDIR "/init.d/ntpd"
-#define NTPRESTARTARGS      "--nodeps", "--quiet", "conditionalrestart"
+#define NTPSERVICE          "/etc/rc.d/init.d/ntpd"
+#define NTPRESTARTARGS      "condrestart"
 
-#define OPENNTPFILE         ETCDIR "/ntpd.conf"
-#define OPENNTPSERVICE      ETCDIR "/init.d/ntpd"
-#define OPENNTPRESTARTARGS  "--nodeps", "--quiet", "conditionalrestart"
+#define OPENNTPFILE         "/etc/ntpd/ntpd.conf"
+#define OPENNTPSERVICE      "/etc/rc.d/init.d/ntpd"
+#define OPENNTPRESTARTARGS  "condrestart"
 
 #define DEFAULT_SCRIPT      ETCDIR "/" PACKAGE ".sh"
 
