--- dhcpcd-5.2.5/configure	2010-06-10 14:05:50.000000000 +0200
+++ dhcpcd-5.2.4/configure	2010-06-04 10:34:28.000000000 +0200
@@ -113,26 +102,11 @@
 	# Derive OS from cpu-manufacturer-os-kernel
 	CPU=${TARGET%%-*}
 	REST=${TARGET#*-}
-	if [ "$CPU" != "$REST" ]; then
-		MANU=${REST%%-*}
-		REST=${REST#*-}
-		if [ "$MANU" != "$REST" ]; then
-			OS=${REST%%-*}
-			REST=${REST#*-}
-			if [ "$OS" != "$REST" ]; then
-				KERNEL=${REST%%-*}
-			else
-				# 3 tupple
-				KERNEL=$OS
-				OS=$MANU
-				MANU=
-			fi
-		else
-			# 2 tupple
-			OS=$MANU
-			MANU=
-		fi
-	fi
+	MANU=${REST%%-*}
+	REST=${REST#*-}
+	OS=${REST%%-*}
+	REST=${REST#*-}
+	KERNEL=${REST%%-*}
 fi
