--- ./dhcpconfig.c.org2	Mon Jan  6 12:31:46 2003
+++ ./dhcpconfig.c	Mon Jan  6 12:36:42 2003
@@ -334,7 +334,7 @@
     ((unsigned char *)&DhcpIface.ciaddr)[1],
     ((unsigned char *)&DhcpIface.ciaddr)[2],
     ((unsigned char *)&DhcpIface.ciaddr)[3]);
-  if ( ReplResolvConf )
+  if ( ReplResolvConf  && DhcpOptions.len[dns] )
     {
       resolv_renamed=1+rename(RESOLV_CONF,""RESOLV_CONF".sv");
       f=fopen(RESOLV_CONF,"w");
@@ -406,7 +406,7 @@
       else
 	syslog(LOG_ERR,"dhcpConfig: fopen: %m\n");
     }
-  if ( ReplNTPConf )
+  if ( ReplNTPConf  && DhcpOptions.len[ntpServers] >= 4 )
     {
       ntp_renamed=1+rename(NTP_CONF,""NTP_CONF".sv");
       f=fopen(NTP_CONF,"w");
